pkg/exporter/sumologicexporter/prometheus_formatter.go

```diff
@@ -57,38 +57,90 @@ func newPrometheusFormatter() (prometheusFormatter, error) {
 
 // PrometheusLabels returns all attributes as sanitized prometheus labels string
 func (f *prometheusFormatter) tags2String(attr pdata.AttributeMap, labels pdata.AttributeMap) prometheusTags {
+	attrsPlusLabelsLen := attr.Len() + labels.Len()
+	if attrsPlusLabelsLen == 0 {
+		return ""
+	}
+
 	mergedAttributes := pdata.NewAttributeMap()
+	mergedAttributes.EnsureCapacity(attrsPlusLabelsLen)
+
 	attr.CopyTo(mergedAttributes)
 	labels.Range(func(k string, v pdata.AttributeValue) bool {
 		mergedAttributes.UpsertString(k, v.StringVal())
 		return true
 	})
 	length := mergedAttributes.Len()
 
-	if length == 0 {
-		return ""
-	}
-
 	returnValue := make([]string, 0, length)
 	mergedAttributes.Range(func(k string, v pdata.AttributeValue) bool {
+		key := f.sanitizeKeyBytes([]byte(k))
+		value := f.sanitizeValue(v.AsString())
+
 		returnValue = append(
 			returnValue,
-			fmt.Sprintf(
-				`%s="%s"`,
-				f.sanitizeKey(k),
-				f.sanitizeValue(v.AsString()),
-			),
+			formatKeyValuePair(key, value),
 		)
 		return true
 	})
 
-	return prometheusTags(fmt.Sprintf("{%s}", strings.Join(returnValue, ",")))
+	return prometheusTags(stringsJoinAndSurround(returnValue, ',', '{', '}'))
+}
+
+func formatKeyValuePair(key []byte, value string) string {
+	// Use strings.Builder and not fmt.Sprintf as it uses significantly less
+	// allocations.
+	sb := strings.Builder{}
+	// We preallocate space for key, value, equal sign and quotes.
+	sb.Grow(len(key) + len(value) + 1 + 2)
```

Comment:
I think that it would be more readable to not use hardcoded values in this case
`const equalSign string = "="`
 `const quoteSign string = "\""`
 `sb.Grow(len(key) + len(value) + len(equalSign) + len(quoteSign) + len(quoteSign)`
This still does not guarantee consistency of allocated buffer and real space used to contain all characters.
One example of fixing it, would be, hiding string builder under higher level builder api with at least two functions for one allocating space and second for building string which would check (at runtime) consistency between allocated space and really used one.

Reply:
What would be the outcome of that check in your view @pdelewski? Let's say you try to write more data to the buffer than what is allocated, StringBuilder will automatically allocate more space. Should we then panic, print a warning, or something else, instead?

Reply:
It depends if that preallocation is critical for us and if that's the case we can return error or panic

Reply:
We can benchmark this with and without the preallocation if we really want to find the performance difference between the 2 approaches but @swiatekm-sumo has a point that `strings.Builder` does re-allocate when needed and it's its feature so there's really no point to work around that I'd say.

To address your suggestion @pdelewski we could make those `const`s but then what type should they be? If we stick to `rune`s then they are always of length 1 and **you cannot take `rune`'s length** i.e.

```go
r := '['
l := len(r) // <- this won't compile
```

In that case we'd need to resort to strings but then we'd have to measure if it makes sense from performance standpoint (probably it does).

Reply:
Do we need to use runes? Are they better, faster in this particular case?

Reply:
It looks like it's basically the same:

This PR:

```
goos: darwin
goarch: amd64
pkg: github.com/SumoLogic/sumologic-otel-collector/pkg/exporter/sumologicexporter
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark_PrometheusFormatter_Metric2String
Benchmark_PrometheusFormatter_Metric2String-16             22712             51177 ns/op           15821 B/op        512 allocs/op
Benchmark_PrometheusFormatter_Metric2String-16             23038             50360 ns/op           15791 B/op        512 allocs/op
Benchmark_PrometheusFormatter_Metric2String-16             23560             51604 ns/op           15810 B/op        512 allocs/op
PASS
ok      github.com/SumoLogic/sumologic-otel-collector/pkg/exporter/sumologicexporter    5.578s
```

---

This approach:

```go
	sb.WriteString("=")
	sb.WriteString(`"`)
	sb.WriteString(value)
	sb.WriteString(`"`)
```

```
goos: darwin
goarch: amd64
pkg: github.com/SumoLogic/sumologic-otel-collector/pkg/exporter/sumologicexporter
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark_PrometheusFormatter_Metric2String
Benchmark_PrometheusFormatter_Metric2String-16             23334             50641 ns/op           15804 B/op        512 allocs/op
Benchmark_PrometheusFormatter_Metric2String-16             24022             51114 ns/op           15809 B/op        512 allocs/op
Benchmark_PrometheusFormatter_Metric2String-16             24036             49831 ns/op           15811 B/op        512 allocs/op
```

Reply:
I would stick to string if you don't mind (it's simpler to read as we don't need to mix different types and we avoid questions why runes are there)

Reply:
Please post your review then :)  👍 / 👎

Reply:
:)